src/object/obj_class.c

```diff
@@ -234,9 +234,24 @@ daos_oclass_grp_nr(struct daos_oclass_attr *oc_attr, struct daos_obj_md *md)
 	return oc_attr->ca_grp_nr;
 }
 
+/**
+ * To honor RF setting during failure cases, let's reserve RF
+ * groups, so if some targets fail, there will be enough replacement
+ * targets to rebuild, so to avoid putting multiple shards in the same
+ * domain, which may break the RF setting.
+ *
+ * Though let's keep reserve targets to be less than 30% of the total
+ * targets.
+ */
+static uint32_t
+reserve_grp_by_rf(uint32_t target_nr, uint32_t grp_size, uint32_t rf)
+{
+	return min(((target_nr * 3) / 10) / grp_size, rf);
```

Comment:
> im still not convinced that an object with EC or RP, in a container with RF0 is not supported here..
> it sounds like we are telling users that containers with RF0 do not support any object with redundancy, however it should.

I think this is what RF means, once the failure reach the RF, the container will turns to inaccessible.  There are no way for user to know if there are data lost in their container, unless we do object scan etc, but this is not supported in the current version.

So object class have to work together with RF to achieve the data redundancy in the current version.


> same case applies for an object with EC_xP2 or RP3 in a container with RF1 .. we don't reserve enough in this case here either right?

Yes, reservation is by RF,  which is the only one to be used for checking the data lost for now.

Reply:
i agree that this is at least better than what we have now. let's document this properly though for users to understand the limitation of using redundant object class but with RF that is at a lower level of redundancy (this can be a common usecase fwiw).